Enable CPU-hoisting for `stablehlo.dynamic_update_slice`

dmilinkovicTT opened the pull request:
### Ticket
#5527 

### Problem description
`stablehlo.dynamic_update_slice` conversion to TTIR isn't supported. We need to hoist execution of this op to the CPU.

Most of the efforts for supporting this op are focused on enabling integer support in CPU-hoisted funcs.

### What's changed
- Allowed integer types in CPU-hoisted functions
- Appended `linalg.copy` to the end of CPU-hoisted functions if DPS semantics can't be enabled by other means
- Allowed creating `TTIRHoistPass` which would target only certain ops
- Generalized implementation of `TTIRHoistAnalyze` and `TTIRHoistTransform`, so that they receive a predicate which determines whether an op should be hoisted
- Disabled execution of `ElementTypeNormalization` on CPU module
- Fixed lowering of `ttnn.full` to flatbuffer
- Enabled CPU-hoisting stablehlo ops by default

### Checklist
- [x] New/Existing tests provide coverage for changes

vwellsTT reviewed (approved):
I still think it'd be good to add some `builder` tests before merging, FWIW
But otherwise great work, having int support is nice!

azecevicTT commented on lib/Dialect/TTIR/Pipelines/TTIRPipelines.cpp:
```diff
@@ -72,10 +83,17 @@ void createStableHLOToTTIRPipeline(
+    // if any unsupported ops have not been hoisted.
+    auto &devicePm = pm.nest<ttcore::DeviceModuleOp>().nest<mlir::ModuleOp>();
+    passOptions.enablePartialConversion = false;
+    devicePm.addPass(createConvertStableHLOToTTIRPass(passOptions));
```
The natural order here should be `SHLO->TTIR` and then cpu fallback. Otherwise, we might end up with a cpu op that could be executed on device. I understand from implementation perpective why you did it this way, one alternative to this would be to change the type of `SHLO->TTIR` from conversion to greedy rewriter, but that would require some non-trivial changes. Other sound option would be to first run `SHLO->TTIR` with `enablePartialConversion=true`, then cpu fallback, and finally `SHLO->TTIR` with `enablePartialConversion=false` (this second run would be pretty much verification run).

azecevicTT commented on lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp:
```diff
@@ -293,33 +309,40 @@ static void hoistOperationToFunction(mlir::Operation *opToHoist,
-// `should_hoist` attribute, but in the future will also tag fall-back ops, etc.
 namespace {
+// Predicate type for determining whether an op should be hoisted.
+using ShouldHoistPredicate = std::function<bool(mlir::Operation *)>;
```
Nit: I find convention of putting trailing `T`, `Ty` or `Type` very useful in cases like this one. `ShouldHoistPredicate` sounds more like a function name than function type name.

azecevicTT commented on lib/Dialect/TTIR/Transforms/HoistCPUOps.cpp:
```diff
@@ -397,23 +433,42 @@ class TTIRHoistTransform
+template <typename... Ops>
+std::unique_ptr<mlir::Pass> createTTIRHoistTransformForOps() {
+  const auto customPredicate = [](mlir::Operation *op) {
+    return (llvm::isa<Ops>(op) || ...);
```
Nit: `llvm::isa<Ops...>(op)` should be the same and it's little more readable IMO

dmilinkovicTT commented:
@vwellsTT thanks for the review!

I've added a silicon test for `dynamic_update_slice`, and I've updated some hoisted op builder tests so that they are parametrized with both f32 and i32 types.

I've also adjusted the ptr type in `WrappedTensor`, as we've discussed offline
